### PR TITLE
feat(express-relay): Support websocket in the js sdk

### PR DESCRIPTION
**Pull request by m30m**

### CODE REVIEW

File: express_relay/sdk/js/src/index.ts [hunk 1]
```diff
@@ -118,13 +120,140 @@ function checkTokenQty(token: { contract: string; amount: string }): TokenQty {
...
+  ) => Promise<void>;
 
-  constructor(clientOptions?: ClientOptions) {
+  constructor(clientOptions: ClientOptions) {
     this.clientOptions = clientOptions;
   }
 
+  private connectWebsocket() {
+    this.websocket = new WebSocket(
+      this.clientOptions.baseUrl.replace("http", "ws") + "/v1/ws"
```
Comment (anihamde) [inline express_relay/sdk/js/src/index.ts:143]: why the potential double occurrence of "ws"?

File: express_relay/sdk/js/src/index.ts [hunk 2]
```diff
@@ -118,13 +120,140 @@ function checkTokenQty(token: { contract: string; amount: string }): TokenQty {
...
+        }
+      } else if ("error" in message) {
+        // Can not route error messages to the callback router as they don't have an id
+        console.error(message.error);
+      }
+    });
+  }
+
+  private convertOpportunity(
+    opportunity: components["schemas"]["OpportunityParamsWithMetadata"]
```
Comment (anihamde) [inline express_relay/sdk/js/src/index.ts:171]: maybe some comments here? also a check for params version?

**Review by anihamde — APPROVED** [review]

File: express_relay/sdk/js/src/index.ts [hunk 3]
```diff
@@ -118,13 +120,157 @@ function checkTokenQty(token: { contract: string; amount: string }): TokenQty {
...
+          callback(message);
+          delete this.callbackRouter[message.id];
+        }
+      } else if ("type" in message && message.type === "new_opportunity") {
+        if (this.websocketOpportunityCallback !== undefined) {
+          const convertedOpportunity = this.convertOpportunity(
+            message.opportunity
+          );
+          if (convertedOpportunity !== undefined) {
+            await this.websocketOpportunityCallback(convertedOpportunity);
```
Comment (ali-bahjati) [inline express_relay/sdk/js/src/index.ts:166]: What happens if this is undefined? Maybe it be nice to take this lambda function in constructor (if it's critical for the client)

File: express_relay/sdk/js/src/index.ts [hunk 4]
```diff
@@ -118,13 +120,157 @@ function checkTokenQty(token: { contract: string; amount: string }): TokenQty {
...
+
+  async sendWebsocketMessage(
+    msg: components["schemas"]["ClientMessage"]
+  ): Promise<void> {
+    const msg_with_id: components["schemas"]["ClientRequest"] = {
+      ...msg,
+      id: (this.idCounter++).toString(),
+    };
+    return new Promise((resolve, reject) => {
+      this.callbackRouter[msg_with_id.id] = (response) => {
```
Comment (ali-bahjati) [inline express_relay/sdk/js/src/index.ts:250]: i recommend setting a timeout here (to handle network errors better)

**Review by ali-bahjati — APPROVED** [review]

**Review by ali-bahjati — APPROVED** [review]
Nice! I left some comments to improve the code.
I think we need to think more about error-handling in cases of failure and reconnections that can be addressed later.

File: express_relay/sdk/js/src/index.ts (hunk 3 above)
Comment (m30m) [reply express_relay/sdk/js/src/index.ts:166]: They will see the logs about this. I would rather improve this when we get some feedbacks from searchers

**Pull request merged**